platforms/cpuinfo_linux.go

```diff
@@ -0,0 +1,88 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package platforms
+
+import (
+	"bytes"
+	"runtime"
+	"strings"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"golang.org/x/sys/unix"
+)
+
+// getMachineArch retrieves the machine architecture through system call
+// same result from uname -m running the terminal
+// This function is favored to the previous method of getting variant information from /proc/cpuinfo
+// because /proc/cpuinfo is not accurate in user-mode emulation environment
+func getMachineArch() (string, error) {
+	var uname unix.Utsname
+	err := unix.Uname(&uname)
+	if err != nil {
+		return "", err
+	}
+
+	arch := string(uname.Machine[:bytes.IndexByte(uname.Machine[:], 0)])
+
+	return arch, nil
+}
+
+// getCPUVariant returns cpu variant for ARM
+// return empty string in case of error
+func getCPUVariant() string {
+
+	if !isArmArch(runtime.GOARCH) {
+		log.L.Errorf("getCPUVariant for ARCH %s: %v", runtime.GOARCH, errdefs.ErrNotImplemented)
+	}
+
+	var variant string
+
+	arch, err := getMachineArch()
+	if err != nil {
+		log.L.WithError(err).Error("failure getting machine architecture")
+		return ""
+	}
+
+	arch = strings.ToLower(arch)
+
+	if arch == "aarch64" {
+		variant = "v8"
```

Comment:
This is one place the old implementation was probably superior - on `arm64`, it would (correctly) get `8` from `/proc/cpuinfo`, which is more likely to continue working in the future and get `9`, etc.  Perhaps we should only fall back to this naive `uname -m`-based implementation when the `/proc/cpuinfo` method fails to determine the variant?

(Especially since the implementation of `amd64` variants is likely going to have to look at features from `/proc/cpuinfo` to determine the appropriate variant.)

Reply:
Thanks for your comments! @tianon.
Actually my first version of the PR was using uname as a fall back.  After getting some feedback from community, I was convinced that uname approach is more clear and most accurate(no need to handle edge case and it's a realtime system call which can reflect user-mode emulation like issues included in the PR). 
We need to handle the special case for aarch64 as it's not following the same armvX convention and the old approach also had a case check for that.  As for for variant 9, we need to add another check for that no matter which approach we choose now. 
The getCPUVariant function is only supposed to run on ARM platforms, we don't need the variant info for amd64.